### ISSUE TITLE
Terraform does not detect or correct drift when an okta_app_saml resource is assigned to a different Authentication Policy in Okta  UI

### ISSUE THREAD

**Issue by swu-etsy** [okta/terraform-provider-okta#1460]
### Community Note

- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
- If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

v1.3.9

### Affected Resource(s)

- okta_app_saml

### Terraform Configuration Files

```hcl
resource "okta_app_saml" "org2org" {
  label             = "Org2Org"
  preconfigured_app = "okta_org2org"
  status            = "ACTIVE"

  hide_ios = true
  hide_web = true

  authentication_policy = {POLICY_ID_GOES_HERE}

  lifecycle {
    ignore_changes = [groups]
  }

}
```

### Expected Behavior

When, in the Admin Console GUI, an app managed by `okta_app_saml` in Terraform is assigned to a different authentication policy than the one specified in the resource's `authentication_policy` argument:

`terraform plan` should show drift between the app's real policy value in Okta and the policy value specified in the `authentication_policy` argument in the Terraform configuration.

`terraform apply` should restore the app's policy in Okta to the one that is specified in the `authentication_policy` argument in the Terraform configuration.

### Actual Behavior
When, in the Admin Console GUI, an app managed by the `okta_app_saml` is assigned to a different authentication policy than the one specified in the resource's `authentication_policy` property:
`terraform plan` shows no drift.
`terraform apply` does not update the app's authentication policy to the one specified in the Terraform resource.

### Steps to Reproduce

1. Create an `okta_app_saml` resource block in Terraform.
2. In the resource's `authentication_policy` argument, pass in the policy id of an existing authentication policy.
3. `terraform apply`.
4. Log into Okta's Admin Console GUI. Under Security -> Authentication Policies, and assign the app created in step 1 to a different authentication policy from the one specified in Step 2 above.
5. `terraform plan` -> Observe no drift in the app's `authentication_policy` value.
6. `terraform apply` -> Observe that the app is still assigned to the policy it was assigned to in Step 4 above.

**Comment by monde** [comment]
@swu-etsy the resource has changed outside of terraform, so what is really needed is a refresh, no? `terraform apply -refresh-only` https://developer.hashicorp.com/terraform/cli/commands/refresh

**Comment by monde** [comment]
Seems like they are talking about the state of the art in refresh here https://support.hashicorp.com/hc/en-us/articles/4405950960147-New-Feature-Objects-have-changed-outside-of-Terraform-

**Comment by ericnorris** [comment]
Hey @monde, I'm not sure I follow - as I understand it, Terraform executes a refresh during a `plan` and `apply` cycle, which seems to be confirmed by both your first and second links ("You shouldn't typically need to use this command, because Terraform automatically performs the same refreshing actions as a part of creating a plan in both the terraform plan and terraform apply commands." and "This is a task Terraform has done historically by design during refresh silently in the background.").

In step 5 of @swu-etsy's reproduction steps, no drift is detected during the `plan`. I believe that running refresh alone would have the same issue, so my understanding is that there is a bug in the Okta provider as it's not providing enough information to the Terraform Core for it to know there is drift.

**Comment by ericnorris** [comment]
Looking into this a bit more, as far as I can tell `d.Set("authentication_policy", ...)` is never called in `resourceAppSamlRead`:

https://github.com/okta/terraform-provider-okta/blob/4c35c37d6f00f614aaf696c6ef022277b29d60cf/okta/resource_okta_app_saml.go#L478-L560

That seems to confirm that Terraform Core isn't aware of a change to the `authentication_policy` from the remote source.

**Comment by monde** [comment]
Right, I was just looking at the same code @ericnorris.

I need to look into Terraform best practices here. If I recall correctly, an original sin of the Okta Terraform provider with some resources is that more than one API call has to be made to represent a resource. Effectively some resources are a composite of a number of API calls such as this example. That is a bit of an aside. In this case the need to make multiple API calls to get all of the attributes of a resource is an API architecture design choice made in the Okta API's development.

So, to @ericnorris 's point, even if it is bad form for a resource to have to make multiple API calls to infer state, this is necessitated by the Okta API not providing all information for the saml app in a call to `GET /api/v1/apps/{appId}`.

So this is bug on the part of `okta_app_saml`.

Okta Internal Reference: https://oktainc.atlassian.net/browse/OKTA-580817

**Comment by ericnorris** [comment]
Thanks for the fast response @monde!

> If I recall correctly, an original sin of the Okta Terraform provider with some resources is that more than one API call has to be made to represent a resource.

I'm not a Terraform provider developer myself, but I'd be surprised if "one API call per resource / data source" was a hard rule. The Google provider, which I consider to be pretty exemplary, needs to make multiple API calls for a number of objects. Take for example the `google_projects` data source (https://github.com/hashicorp/terraform-provider-google/blob/main/google/data_source_google_projects.go), which iterates over a list of projects via API calls per page; or 
 the `google_compute_instance` resource (https://github.com/hashicorp/terraform-provider-google/blob/main/google/resource_compute_instance.go), which makes at least two calls: one to start creating the instance, and another to poll whether the instance has finished creating.

I wonder if maybe you were thinking of "Resources should represent a single API object" from https://developer.hashicorp.com/terraform/plugin/hashicorp-provider-design-principles? I agree that Okta's API does perhaps make things difficult here, but I think that ideally an app would be an API object.

**Comment by slichtenthal** [comment]
I am digging into this a little and it actually does not seem like another API call is necessary.  When an API call to` GET /api/v1/apps/{appId}` is made, one of the properties returned is `_links`.  Within that is a key called `accessPolicy`.  And while I have not tested for a huge amount of apps, in every app I have checked so far, this key is there.  So similarly to how `logo_url` and `embed_url` is returned, it should be possible to return `access_policy_url` without making another call.  I did test and changing this policy does in fact update the access_policy_url properly, which should cause drift.

**Comment by monde** [comment]
Sharp eyes @slichtenthal , links have not been my favorite thing to rely on for concrete data but there is a links helper pluck info out of them.

- https://github.com/okta/terraform-provider-okta/blob/master/okta/links.go#L3-L26
- https://github.com/okta/terraform-provider-okta/blob/master/okta/resource_okta_link_value.go#L88

**Comment by slichtenthal** [comment]
Thanks for pointing out these functions!

I agree that going through links is not ideal, especially since the API documentation doesn't seem to provide a solid contract of what values are within the links section, which means accessPolicy could, in theory disappear.  

Knowing that, do do you think it's best to parse the Id out of the URL and call the key `authentication_policy`, or simply put the URL as is into a key called `access_policy_url`?

Looking at previous patterns, it seems like values from the links section are always returned as-is, for example: 
https://github.com/etsy/terraform-provider-okta/blob/fb7c2494ff5d8d593ae90a99e23e37bec3defda1/okta/resource_okta_app_saml.go#L490:L491

I agree this is different as it's impossible to return a logo file and the embed_url is the literal value.

**Comment by monde** [comment]
fyi, I'm messing around with a PR for this and have decided reading the authentication policy id (Okta internals named these access policies) off the links is the way keep keep track of changes.

```go
	accessPolicy := linksValue(app.Links, "accessPolicy", "href")
	if accessPolicy != "" {
		d.Set("authentication_policy", path.Base(accessPolicy))
	}
```

**Comment by monde** [comment]
Released in [v3.43.0](https://github.com/okta/terraform-provider-okta/releases/tag/v3.43.0)
